sdk/dapp-kit/src/hooks/useSuiClientInfiniteQuery.ts

```diff
@@ -0,0 +1,67 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+import type { UseInfiniteQueryOptions } from '@tanstack/react-query';
+import { useInfiniteQuery } from '@tanstack/react-query';
+import { useSuiClientContext } from './useSuiClient.js';
+import type { SuiClient } from '@mysten/sui.js/client';
+
+interface PaginatedResult {
+	data?: unknown;
+	nextCursor?: unknown;
+	hasNextPage: boolean;
+}
+export type SuiRpcPaginatedMethodName = {
+	[K in keyof SuiClient]: SuiClient[K] extends (input: any) => Promise<PaginatedResult> ? K : never;
+}[keyof SuiClient];
+
+export type SuiRpcPaginatedMethods = {
+	[K in SuiRpcPaginatedMethodName]: SuiClient[K] extends (input: infer P) => Promise<{
+		data?: infer R;
+		nextCursor?: infer Cursor | null;
+		hasNextPage: boolean;
+	}>
+		? {
+				name: K;
+				result: R;
+				params: P;
+				cursor: Cursor;
+		  }
+		: never;
+};
+
+export type UseSuiClientInfiniteQueryOptions<T extends keyof SuiRpcPaginatedMethods> = Omit<
+	UseInfiniteQueryOptions<
+		SuiRpcPaginatedMethods[T]['result'],
+		unknown,
+		SuiRpcPaginatedMethods[T]['result'],
+		SuiRpcPaginatedMethods[T]['result'],
+		unknown[]
+	>,
+	'queryFn'
+>;
+
+export function useSuiClientInfiniteQuery<T extends keyof SuiRpcPaginatedMethods>(
+	{
+		method,
+		params,
+	}: {
+		method: T;
+		params: SuiRpcPaginatedMethods[T]['params'];
+	},
+	{ queryKey, enabled = !!params, ...options }: UseSuiClientInfiniteQueryOptions<T> = {},
+) {
+	const suiContext = useSuiClientContext();
+
+	return useInfiniteQuery({
+		...options,
+		queryKey: [suiContext.network, method, params],
+		enabled,
+		queryFn: async () => {
+			return await suiContext.client[method](params as never);
```

Comment:
out of curiosity, how does this `as never` cast work?

Reply:
because this is generic, method could be any method, and params could be any params, and typescript doesn't know that its the right params for the current method here. 

`never` is assignable to anything, so it's kinda like doing as `any`, but any is generally more hacky.

The way I think about it is something like:  when typescript checks if a type is assigable to another type.  If the type is a union, each type in the union is checked for assignability.  `never` is basically an empty union, so checking if an empty union is assignable to a type, so its similar to iterating over an empty list and checking that each item in the list is assignable.

Reply:
thank you for the explainer!